File: test/test.NativeMessagingUpstreamTransport.ts
import assert from "node:assert/strict";
import { once } from "node:events";
import { existsSync } from "node:fs";
import net from "node:net";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { test } from "vitest";

import { NativeMessagingUpstreamTransport } from "../bridge/NativeMessagingUpstreamTransport.js";
import { ModCDPClient } from "../client/js/ModCDPClient.js";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const EXTENSION_PATH = path.resolve(HERE, "..", "dist", "extension");
const nativeHostName = (label: string) => `com.modcdp.test.${label}.${process.pid}`;

test("nativemessaging upstream config owns manifest, host, wait timeout, loopback, and injector config", async () => {
  const transport = new NativeMessagingUpstreamTransport({
    manifest_path: "/tmp/modcdp-native-host.json",
    manifest_paths: ["/tmp/modcdp-native-host-extra.json"],
    host_name: "com.modcdp.test",
    extension_id: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
    wait_timeout_ms: 10,
  });
  assert.deepEqual(transport.getInjectorConfig(), { native_host_name: "com.modcdp.test" });
  assert.deepEqual(transport.getServerConfig(), {});
  assert.equal(
    transport.update({
      ws_url: "ws://127.0.0.1:9222/devtools/browser/test",
      manifest_paths: [],
      native_host_name: "com.modcdp.updated",
      wait_timeout_ms: 5,
    }),
    transport,
  );
  assert.deepEqual(transport.getServerConfig(), {
    loopback_cdp_url: "ws://127.0.0.1:9222/devtools/browser/test",
  });
  assert.deepEqual(transport.getInjectorConfig(), { native_host_name: "com.modcdp.updated" });
  assert.equal(
    (transport as unknown as { include_default_manifest_paths: boolean }).include_default_manifest_paths,
    false,
  );
  transport.update({ manifest_path: null });
  assert.equal(
    (transport as unknown as { include_default_manifest_paths: boolean }).include_default_manifest_paths,
    true,
  );
  transport.update({ user_data_dir: "/tmp/modcdp-profile-one" });
  transport.update({ user_data_dir: "/tmp/modcdp-profile-one" });
  transport.update({ user_data_dir: "/tmp/modcdp-profile-two" });
  assert.deepEqual((transport as unknown as { manifest_paths: string[] }).manifest_paths, [
    path.join("/tmp/modcdp-profile-two", "NativeMessagingHosts", "com.modcdp.updated.json"),
    path.join("/tmp/modcdp-profile-two", "Default", "NativeMessagingHosts", "com.modcdp.updated.json"),
  ]);
  await assert.rejects(
    () => transport.waitForPeer(),
    /Timed out waiting 5ms for native messaging host com\.modcdp\.updated/,
  );
});

test("nativemessaging upstream close rejects pending peer waits", async () => {
  const transport = new NativeMessagingUpstreamTransport({
    host_name: "com.modcdp.close",
    wait_timeout_ms: 5_000,
  });
  const pending = transport.waitForPeer();

  await transport.close();

  await assert.rejects(
    () => pending,
    /Native messaging transport for com\.modcdp\.close closed before a peer connected/,
  );
});

test("nativemessaging upstream close resets peer wait state", async () => {
  const host_name = nativeHostName("close-reset");
  const transport = new NativeMessagingUpstreamTransport({ host_name, wait_timeout_ms: 5 });
  await transport.connect();
  const port = Number(new URL(transport.url).port);
  const peer = net.createConnection({ host: "127.0.0.1", port });
  await once(peer, "connect");

  try {
    await transport.waitForPeer();
    await transport.close();

    await assert.rejects(() => transport.waitForPeer(), {
      message: `Timed out waiting 5ms for native messaging host ${host_name}.`,
    });
  } finally {
    peer.destroy();
    await transport.close();
  }
});

test("nativemessaging upstream waits again after a peer disconnects", async () => {
  const host_name = nativeHostName("disconnect-reset");
  const transport = new NativeMessagingUpstreamTransport({ host_name, wait_timeout_ms: 5 });
  await transport.connect();
  const port = Number(new URL(transport.url).port);
  const peer = net.createConnection({ host: "127.0.0.1", port });
  await once(peer, "connect");

  try {
    await transport.waitForPeer();
    peer.destroy();
    await waitFor(() => (transport as unknown as { socket: unknown | null }).socket === null);

    await assert.rejects(() => transport.waitForPeer(), {
      message: `Timed out waiting 5ms for native messaging host ${host_name}.`,
    });
  } finally {
    peer.destroy();
    await transport.close();
  }
});

test("nativemessaging upstream accepts a replacement peer after disconnect", async () => {
  const host_name = nativeHostName("replacement");
  const transport = new NativeMessagingUpstreamTransport({ host_name, wait_timeout_ms: 500 });
  await transport.connect();
  const port = Number(new URL(transport.url).port);
  const first_peer = net.createConnection({ host: "127.0.0.1", port });
  await once(first_peer, "connect");

  try {
    await transport.waitForPeer();
    first_peer.destroy();
    await waitFor(() => (transport as unknown as { socket: unknown | null }).socket === null);

    const second_peer = net.createConnection({ host: "127.0.0.1", port });
    await once(second_peer, "connect");
    try {
      await transport.waitForPeer();
    } finally {
      second_peer.destroy();
    }
  } finally {
    first_peer.destroy();
    await transport.close();
  }
});

test("nativemessaging upstream installs the launch-profile native host manifest and connects to a real extension", async () => {
  const host_name = nativeHostName("client");
  const native_client = new ModCDPClient({
    launch: {
      mode: "local",
      options: { headless: true, sandbox: process.platform !== "linux" },
    },
    upstream: { mode: "nativemessaging", nativemessaging_host_name: host_name },
    extension: {
      mode: "auto",
      path: EXTENSION_PATH,
      service_worker_url_suffixes: ["/modcdp/service_worker.js"],
      trust_service_worker_target: true,
    },
    server: {
      routes: { "*.*": "loopback_cdp" },
    },
  });

  try {
    await native_client.connect();
    assert.equal(native_client.transport?.mode, "nativemessaging");
    assert.equal(native_client.upstream_endpoint_kind, "modcdp_server");
    assert.match(native_client.transport?.url ?? "", /^native:\/\/.+@127\.0\.0\.1:\d+$/);
    assert.equal(native_client.transport?.url?.startsWith(`native://${host_name}@`), true);
    assert.equal(
      existsSync(path.join(native_client._launched?.profile_dir ?? "", "NativeMessagingHosts", `${host_name}.json`)),
      true,
    );
    const version = (await native_client.send("Browser.getVersion")) as Record<string, unknown>;
    assert.equal(typeof version.product, "string");
    await new Promise((resolve) => setTimeout(resolve, 1_500));
    const second_version = (await native_client.send("Browser.getVersion")) as Record<string, unknown>;
    assert.equal(typeof second_version.product, "string");
  } finally {
    await native_client.close();
  }
}, 90_000);

async function waitFor(predicate: () => boolean, timeout_ms = 2_000) {
  const deadline = Date.now() + timeout_ms;
  while (Date.now() < deadline) {
    if (predicate()) return;
    await new Promise((resolve) => setTimeout(resolve, 20));
  }
  throw new Error("Timed out waiting for condition");
}

File: client/python/tests/test_NativeMessagingUpstreamTransport.py
from __future__ import annotations

import sys
import unittest
from pathlib import Path

from modcdp import ModCDPClient
from modcdp.NativeMessagingUpstreamTransport import DEFAULT_NATIVE_MESSAGING_HOST_NAME, NativeMessagingUpstreamTransport


@unittest.skipIf(sys.platform.startswith("win"), "native messaging profile manifest path is not implemented on Windows")
class NativeMessagingUpstreamTransportTests(unittest.TestCase):
    def test_config_owns_manifest_host_wait_timeout_loopback_and_injector_config(self) -> None:
        transport = NativeMessagingUpstreamTransport(
            {
                "manifest_path": "/tmp/modcdp-native-host.json",
                "manifest_paths": ["/tmp/modcdp-native-host-extra.json"],
                "host_name": "com.modcdp.test",
                "extension_id": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
                "wait_timeout_ms": 10,
            }
        )
        self.assertEqual(transport.getInjectorConfig(), {"native_host_name": "com.modcdp.test"})
        self.assertEqual(transport.getServerConfig(), {})
        self.assertIs(
            transport.update(
                {
                    "ws_url": "ws://127.0.0.1:9222/devtools/browser/test",
                    "manifest_paths": [],
                    "native_host_name": "com.modcdp.updated",
                    "wait_timeout_ms": 5,
                }
            ),
            transport,
        )
        self.assertEqual(
            transport.getServerConfig(),
            {"loopback_cdp_url": "ws://127.0.0.1:9222/devtools/browser/test"},
        )
        self.assertEqual(transport.getInjectorConfig(), {"native_host_name": "com.modcdp.updated"})
        self.assertFalse(transport.include_default_manifest_paths)
        transport.update({"manifest_path": None})
        self.assertTrue(transport.include_default_manifest_paths)
        with self.assertRaisesRegex(RuntimeError, r"Timed out waiting 5ms for native messaging host com\.modcdp\.updated"):
            transport.waitForPeer()

    def test_installs_launch_profile_native_host_manifest_and_connects_to_real_extension(self) -> None:
        cdp = ModCDPClient(
            launch={"mode": "local", "options": {"headless": True, "sandbox": False}},
            upstream={"mode": "nativemessaging"},
            extension={
                "mode": "auto",
                "service_worker_url_suffixes": ["/modcdp/service_worker.js"],
                "trust_service_worker_target": True,
            },
            server={"routes": {"*.*": "loopback_cdp"}},
        )

        try:
            cdp.connect()
            self.assertEqual(cdp.transport.mode if cdp.transport else None, "nativemessaging")
            self.assertEqual(cdp.upstream_endpoint_kind, "modcdp_server")
            transport_url = cdp.transport.url if cdp.transport and cdp.transport.url else ""
            self.assertRegex(transport_url, r"^native://com\.modcdp\.bridge@127\.0\.0\.1:\d+$")
            profile_dir = cdp._launched_browser.get("profile_dir") if cdp._launched_browser else ""
            self.assertTrue(
                (Path(profile_dir) / "NativeMessagingHosts" / f"{DEFAULT_NATIVE_MESSAGING_HOST_NAME}.json").exists()
            )
            version = cdp.send("Browser.getVersion")
            self.assertIsInstance(version["product"], str)
        finally:
            cdp.close()


if __name__ == "__main__":
    unittest.main()
